src/checkers/inference/solver/backend/maxsat/MaxSatSolver.java

```diff
@@ -115,14 +109,59 @@ public MaxSatSolver(SolverEnvironment solverEnvironment, Collection<Slot> slots,
                 // PrintUtils.printResult(result);
             } else {
                 System.out.println("Not solvable!");
+                result = null;
             }
 
-        } catch (Throwable e) {
-            e.printStackTrace();
+        } catch (ContradictionException e) {
+            InferenceMain.getInstance().logger.warning("Contradiction exceptin: ");
+            // This case indicates that constraints are not solvable, too. This is normal so continue
+            // execution and let solver strategy to explain why there is no solution
+            result = null;
+        } catch (Exception e) {
+            ErrorReporter.errorAbort("Unexpected error occurred!", e);
         }
         return result;
     }
 
+    /**
+     * Convert constraints to list of VecInt.
+     */
+    @Override
+    public void encodeAllConstraints() {
+        for (Constraint constraint : constraints) {
+            collectVarSlots(constraint);
+            VecInt[] encoding = constraint.serialize(formatTranslator);
+            if (encoding == null) {
+                InferenceMain.getInstance().logger.warning(getClass()
+                        + "doesn't support encoding constraint: " + constraint
+                        + "of class: " + constraint.getClass());
+                continue;
+            }
+            for (VecInt res : encoding) {
+                if (res != null && res.size() != 0) {
+                    if (constraint instanceof PreferenceConstraint) {
+                        softClauses.add(res);
+                    } else {
+                        hardClauses.add(res);
+                        //System.out.println("Generated hard clause: " + res);
+                        // Add here to avoid second round of iteration over constraints if there's no solution
+                        unsatisfiableConstraintExplainer.addVecIntToConstraintMapping(res, constraint);
```

Comment:
@topnessman  Mier, pre-caching constraints for `unsatisfiableConstraintExplainer` seems save a round, but it actually means for normal cases we still need to add these constraints to `unsatisfiableConstraintExplainer`.

Therefore, maybe apply a lazy evaluation of constraints for `unsatisfiableConstraintExplainer` might be a better choice, as:

1. We save the effort of adding constraints to `unsatisfiableConstraintExplainer` for normal cases, ie when CFI has solution, which should be the most cases.

2. We make our code clearer --- ie. at first glance adding constraints to an `unsatisfiableConstraintExplainer` during "encoding" phase seems a bit confusing (the comment clarifies why we have code here, but still, self-explained code is better than code needs comment).

Reply:
Agree. It seems weird to also store `VecInt` to `Constraint` mapping when encoding all `Constraint`s.